caddyhttp/fastcgi/fastcgi.go

```diff
@@ -73,9 +82,30 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error)
 
 			// Connect to FastCGI gateway
 			network, address := rule.parseAddress()
-			fcgiBackend, err := Dial(network, address)
-			if err != nil {
-				return http.StatusBadGateway, err
+
+			var fcgiBackend *FCGIClient
+			reuse := false
+			if usePersistentFcgiConnections {
+				metaMutex.Lock()
+				// re use connection, if possible
+				if _, ok := poorMansSerialisation[network+address]; !ok {
+					poorMansSerialisation[network+address] = new(sync.Mutex)
+				}
+				metaMutex.Unlock()
+
+				poorMansSerialisation[network+address].Lock()
+				defer poorMansSerialisation[network+address].Unlock()
```

Comment:
Well, you need to protect the map accesses, not merely the values inside it. Whenever you access a map concurrently, it must be synced with mutex.